server/server.go

```diff
@@ -65,59 +227,180 @@ func readFormData(r *http.Request) string {
 	fmt.Printf("File Size: %+v\n", handler.Size)
 	fmt.Printf("MIME Header: %+v\n", http.DetectContentType(fileHeader))
 
+	// Decompress the file and return its response.
 	return decompressFile(file, fileHeader, handler)
 }
 
-// Reads all files from the uploaded compressed file
+// decompressFile reads and stores all files from the uploaded compressed file.
 func decompressFile(file multipart.File, fileHeader []byte, handler *multipart.FileHeader) string {
-	// Based on the type of file compression, read the file.
-	if http.DetectContentType(fileHeader) == "application/octet-stream" {
-		// For tar or tar.gz file.
-		unTarred := tar.NewReader(file)
-		for {
-			header, err := unTarred.Next()
-			if err == io.EOF {
-				// End of tar file.
-				break
-			}
-			if err != nil {
-				responseString := `"UnTar Error":"Error in un-tarring uploaded file"`
-				log.Println(err)
-				return responseString
-			}
-			fmt.Println(header.Name)
-		}
-	} else {
-		// For zip file.
+
+	// Read the file based on the type of file compression.
+	assignTestingInfo.RootDir = strings.TrimSuffix(handler.Filename, path.Ext(handler.Filename))
+
+	if http.DetectContentType(fileHeader) == constants.ZipMimeFileType {
+		// Read zip file.
 		unZipped, err := zip.NewReader(file, handler.Size)
 		if err != nil {
+
 			responseString := `"Unzip Error":"Error in unzipping uploaded file"`
-			log.Println(err)
+			log.Println("error in unzipping file", err)
 			return responseString
 		}
-		for _, file := range unZipped.File {
-			fmt.Println(file.Name)
+		return storeUnzippedFiles(unZipped)
+
+	} else if http.DetectContentType(fileHeader) == constants.TarGzMimeFileType {
+		// Read tar.gz file.
+		assignTestingInfo.RootDir = strings.TrimSuffix(assignTestingInfo.RootDir,
+			path.Ext(assignTestingInfo.RootDir))
+		fileReader, err := handler.Open()
+		gZipReader, err := gzip.NewReader(fileReader)
+		if err != nil {
+			responseString := `"Untar Error":"Error in untaring uploaded file"`
+			log.Println("error in untaring file", err)
+			return responseString
 		}
+		unTarred := tar.NewReader(gZipReader)
+		return storeUnTarredFiles(unTarred)
+
+	} else {
+		// Read tar file.
+		var fileReader io.ReadCloser = file
+		unTarred := tar.NewReader(fileReader)
+		return storeUnTarredFiles(unTarred)
 	}
+}
 
+// storeUnTarredFiles stores unTared files to 'assignments' directory.
+func storeUnTarredFiles(unTarred *tar.Reader) string {
```

Comment:
Seems like the same error message is returned regardless of what happens.
It would be better to actually return the error object instead of just the message.

Reply:
That message is actually the response to be sent to the client